src/EditorFeatures/Core/EmbeddedLanguages/Json/JsonEmbeddedBraceMatcher.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Features.EmbeddedLanguages.Json;
+using Microsoft.CodeAnalysis.Features.EmbeddedLanguages.Json.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Editor.EmbeddedLanguages.Json
+{
+    using JsonToken = EmbeddedSyntaxToken<JsonKind>;
+
+    /// <summary>
+    /// Brace matching impl for embedded json strings.
+    /// </summary>
+    internal class JsonEmbeddedBraceMatcher : IBraceMatcher
+    {
+        private readonly EmbeddedLanguageInfo _info;
+
+        public JsonEmbeddedBraceMatcher(EmbeddedLanguageInfo info)
+        {
+            _info = info;
+        }
+
+        public async Task<BraceMatchingResult?> FindBracesAsync(
+            Document document, int position, BraceMatchingOptions options, CancellationToken cancellationToken)
+        {
+            if (!options.HighlightRelatedJsonComponentsUnderCursor)
+                return null;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var detector = JsonLanguageDetector.GetOrCreate(semanticModel.Compilation, _info);
+            var tree = detector.TryParseString(token, semanticModel, cancellationToken);
+            if (tree == null)
+                return null;
+
+            return GetMatchingBraces(tree, position);
+        }
+
+        private static BraceMatchingResult? GetMatchingBraces(JsonTree tree, int position)
+        {
+            var virtualChar = tree.Text.FirstOrNull(vc => vc.Span.Contains(position));
```

Comment:
No binary search or something for this? I can imagine really huge JSON...

Reply:
good point.  i'll add a faster mechanism.

Reply:
done. we're now O(1) for simple strings without escapes, and O(log n) for ones that do ahve them.